datastore-design/go-grpc-demo/storagebackend/timescaledb/findtimeseries.go

```diff
@@ -0,0 +1,127 @@
+package timescaledb
+
+import (
+	"datastore/datastore"
+	"fmt"
+
+	postgis "github.com/cridenour/go-postgis"
+	_ "github.com/lib/pq"
+)
+
+// matchAnyValCond creates the sub-condition of a WHERE clause that checks if string column 'name'
+// matches any of a set of values.
+// Returns the sub-condition to be appended to a WHERE clause.
+func matchAnyValCond(name string, values []string) string {
+	if len(values) == 0 {
+		return ""
+	}
+	inArg := ""
+	for i, value := range values {
+		if i > 0 {
+			inArg += ","
+		}
+		inArg += fmt.Sprintf("'%s'", value)
+	}
+	return fmt.Sprintf(" AND %s IN (%s)", name, inArg)
+}
+
+func equal(p1, p2 *datastore.Point) bool {
+	return (p1.Lat == p2.Lat) && (p1.Lon == p2.Lon)
+}
+
+// insidePolygonCond creates a sub-condition of a WHERE clause that checks if geo point column
+// 'name' is contained in a geo polygon.
+// Returns the sub-condition to be appended to a WHERE clause.
+func insidePolygonCond(name string, polygon0 *datastore.Polygon) (string, error) {
+
+	if polygon0 == nil {
+		// absent polygon => disable filtering (note: "" is equivalent to " AND TRUE")
+		return "", nil
+	}
+
+	points := polygon0.Points
+
+	// validate
+	if len(points) < 3 {
+		return "", fmt.Errorf("polygon must contain at least 3 points; found %d", len(points))
+	}
+	if equal(points[0], points[len(points) - 1]) {
+		return "", fmt.Errorf("polygon endpoints must be different")
+	}
+
+	// construct the polygon ring of the WKT representation
+	// (see https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry;
+	// note that only a single ring is supported for now)
+	polygonRing := ""
+	points = append(points, points[0]) // close polygon
+	for i, point := range points  {
+		if i > 0 {
+			polygonRing += ","
+		}
+		polygonRing += fmt.Sprintf("%f %f", point.Lon, point.Lat)
+	}
+
+	srid := "4326" // spatial reference system ID
+
+	return fmt.Sprintf(
```

Comment:
In this case I don't see any risk of SQL injection since there is no external input of type string, only the lat/lon components, both of which are assumed to be float numbers. The worst that can happen is that you specify a bad polygon, but that's fine.

Reply:
Fair enough, but this assumes type safety is held all the way through.
I remember getting an exception (null pointer, if I remember correctly) in the go datastore by passing in the wrong body type in one of the time-series messages from Python.
I am worried that this could be used to inject something anyway. That's why usually go with "don't interpolate any user input manually".

Reply:
Good point. Better be on the safe side. Update coming up.

Reply:
Pushed fix now.